File: analytics/src/main/java/io/appmetrica/analytics/impl/AppMetricaUncaughtExceptionHandler.java
package io.appmetrica.analytics.impl;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.VisibleForTesting;
import io.appmetrica.analytics.impl.crash.client.AllThreads;
import io.appmetrica.analytics.impl.crash.client.ICrashProcessor;
import io.appmetrica.analytics.impl.crash.utils.CrashedThreadConverter;
import io.appmetrica.analytics.impl.crash.utils.ThreadsStateDumper;
import io.appmetrica.analytics.impl.utils.ProcessDetector;
import io.appmetrica.analytics.logger.internal.DebugLogger;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

public class AppMetricaUncaughtExceptionHandler implements Thread.UncaughtExceptionHandler {
    private static final String TAG = "[MetricaUncaughtExceptionHandler]";

    private static final AtomicBoolean processDying = new AtomicBoolean();

    @NonNull
    private final List<ICrashProcessor> mCrashProcessors;
    @Nullable
    private final Thread.UncaughtExceptionHandler mDefaultHandler;
    @NonNull
    private final ThreadsStateDumper threadsStateDumper;
    @NonNull
    private final ProcessDetector processDetector;
    @NonNull
    private final CrashedThreadConverter crashedThreadConverter;

    public AppMetricaUncaughtExceptionHandler(@Nullable Thread.UncaughtExceptionHandler defaultHandler,
                                              @NonNull List<ICrashProcessor> crashProcessors) {
        this(
                defaultHandler,
                crashProcessors,
                ClientServiceLocator.getInstance().getProcessDetector(),
                new CrashedThreadConverter(),
                new ThreadsStateDumper()
        );
    }

    @VisibleForTesting
    AppMetricaUncaughtExceptionHandler(Thread.UncaughtExceptionHandler defaultHandler,
                                       @NonNull List<ICrashProcessor> crashProcessors,
                                       @NonNull ProcessDetector processDetector,
                                       @NonNull CrashedThreadConverter crashedThreadConverter,
                                       @NonNull ThreadsStateDumper threadsStateDumper) {
        mCrashProcessors = crashProcessors;
        mDefaultHandler = defaultHandler;
        this.processDetector = processDetector;
        this.crashedThreadConverter = crashedThreadConverter;
        this.threadsStateDumper = threadsStateDumper;
    }

    public void uncaughtException(Thread thread, Throwable ex) {
        try {
            DebugLogger.info(TAG, "Process is dying");
            processDying.set(true);
            processUnhandledException(
                    ex,
                    new AllThreads(
                            crashedThreadConverter.apply(thread),
                            threadsStateDumper.getThreadsDumpForCrash(thread),
                            processDetector.getProcessName()
                    )
            );
        } finally {
            if (mDefaultHandler != null) {
                mDefaultHandler.uncaughtException(thread, ex);
            }
        }
    }

    public static boolean isProcessDying() {
        return processDying.get();
    }

    @VisibleForTesting
    void processUnhandledException(@Nullable Throwable originalException,
                                   @NonNull AllThreads allThreads) {
        for (ICrashProcessor crashProcessor : mCrashProcessors) {
            crashProcessor.processCrash(originalException, allThreads);
        }
    }

    @VisibleForTesting(otherwise = VisibleForTesting.NONE)
    @NonNull
    List<ICrashProcessor> getCrashProcessors() {
        return mCrashProcessors;
    }
}

File: analytics/src/main/java/io/appmetrica/analytics/impl/crash/jvm/client/AppMetricaUncaughtExceptionHandler.kt
package io.appmetrica.analytics.impl.crash.jvm.client

import androidx.annotation.VisibleForTesting
import io.appmetrica.analytics.coreutils.internal.logger.LoggerStorage
import io.appmetrica.analytics.impl.ClientServiceLocator
import io.appmetrica.analytics.impl.crash.utils.CrashedThreadConverter
import io.appmetrica.analytics.impl.crash.utils.ThreadsStateDumper
import io.appmetrica.analytics.logger.appmetrica.internal.DebugLogger
import java.util.concurrent.atomic.AtomicBoolean

class AppMetricaUncaughtExceptionHandler(
    private val crashProcessor: ICrashProcessor
) : Thread.UncaughtExceptionHandler {

    private val tag = "[AppMetricaUncaughtExceptionHandler]"

    private val processNameProvider = ClientServiceLocator.getInstance().processNameProvider
    private val crashedThreadConverter = CrashedThreadConverter()
    private val threadsStateDumper = ThreadsStateDumper()

    override fun uncaughtException(thread: Thread, ex: Throwable) {
        try {
            DebugLogger.info(tag, "Process is dying")
            processDying.set(true)
            crashProcessor.processCrash(
                ex,
                AllThreads(
                    crashedThreadConverter.apply(thread),
                    threadsStateDumper.getThreadsDumpForCrash(thread),
                    processNameProvider.getProcessName()
                )
            )
        } catch (e: Throwable) {
            LoggerStorage.getMainPublicOrAnonymousLogger().error(e, e.message)
        }
    }

    companion object {

        private val processDying = AtomicBoolean()

        @JvmStatic
        fun isProcessDying(): Boolean {
            return processDying.get()
        }

        @VisibleForTesting
        fun reset() {
            processDying.set(false)
        }

        @VisibleForTesting
        fun set() {
            processDying.set(true)
        }
    }
}
